Add custom native toolchains to build-script

etcwilde opened the pull request:
This patch adds the ability to specify custom locations to look for build tools.
The main idea is to be able to re-use an already built swift toolchain for building with freestanding configurations.
I'm not sure exactly how the python toolchain pieces integrate with `build-script-impl`.

rdar://92540252

etcwilde committed:
Allow specifying alternative root locations

This patch is intended to enable providing alternative toolchain
locations. This is intended primarily for cross-compilations where I
have a very specific compiler that I want to have building the runtimes
and cannot have the actual host compiler involved at any point during
the build. The provided toolchain must contain all of the necessary
binaries for the build.

etcwilde committed:
Add fallback option

If the tool is not found in the provided path, we should be able to
fallback on the host system if necessary. If fallback is set to true, we
will fall back on using `which` to find the tool.

etcwilde commented:
@swift-ci please test

buttaface commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -594,6 +594,10 @@ def create_argument_parser():
 
+    option('--custom-native-toolchain', store,
+           default=None,
+           help="Specify a custom location to find compilers and tools")
```
Interesting, how is this intended to differ from the existing `--native-{swift,clang,llvm}-tools-path` flags?

gottesmm replied:
@buttaface I think that with time, we can eliminate the usage of that flag in favor of the custom toolchain option. I don't think that we ever specify them as being in different places.

@etcwilde it would be great if you could transition those as well in a subsequent PR. I think it would be pretty easy (just change the build-preset file that specifies this to specify custom native toolchain). It would be a better end product, but if you are worried about time, I am fine with adding a TODO even though I would prefer otherwise.

etcwilde replied:
@buttaface I had looked at using those, and from the docs they seemed like they should have worked.
Maybe I missed it, but I didn't see how those affected the compiler selected for python products (Toolchain::find_tools doesn't seem to consider it at all), and on log inspection, I was picking up the wrong (host, not toolchain) compiler for some of the build products, so I'd figured I didn't fully understand the semantics of the flags. I think longer-term, you may actually want a proper sterile workspace for all of the tools. 

@gottesmm Yeah, I can do that. Like you said, they could come from different places, though they probably should actually be coming from the same place to ensure that the tools are compatible.

buttaface replied:
Those three flags to set paths to prebuilt LLVM, clang, or Swift tools have [been in `build-script-impl` since the open source release](https://github.com/apple/swift/commit/4ac9c80809450e4566946f40bbdb639b2ba745d8#diff-1a5a064af373f742294e3b312a5c8d0ab52f5aef78052490e6005165472cda48R146), to which [I added these Python build-script flags so we could build Python products like SPM too](https://github.com/apple/swift/pull/34437/files#diff-d7765215259ca629c1a1910c1f60768d4e5c9abafd87f624f97f0d3eff33960cR377).

@etcwilde, since those old bash flags were simply bridged over into Python by me, I was unaware of `Toolchain::find_tool` and since all the build products I cared about worked or required other changes anyway, hadn't looked at it. If you want to consolidate these old flags and tie them into the Python build process better, just let us know how you want to proceed, but I think adding yet another flag here on top of the three existing flags would be a mistake.

etcwilde replied:
That's what I had thought from the docs.

Anyway, the llvm flag connects up with `SWIFT_NATIVE_LLVM_TOOLS_PATH` in cmake. That variable only seems to get set, not used, so I think `--native-llvm-tools-path` doesn't actually do anything.

`SWIFT_NATIVE_CLANG_TOOLS_PATH` sets `CMAKE_CXX_COMPILER` and `CMAKE_C_COMPILER` in the stdlib and runtime build, so that does actually have an effect, though is a straightforward.

`SWIFT_NATIVE_SWIFT_TOOLS_PATH` is the only one that is non-trivial to replace from a quick glance, though I suspect it would only take a bit of cleanup to get it to the point of setting `CMAKE_Swift_COMPILER` appropriately.
A quick search has it showing up in localization and the bootstrapping mode on non-darwin platforms. Since the new flag does a walk of the directories to find the tools, the original directories passed to these flags will still work in a migration.

So anyway, I agree, there shouldn't be four flags that are all for effectively the same purpose. I would like to split the adding of the new flag and removal of the old flags into separate PRs though.

buttaface replied:
>  so I think --native-llvm-tools-path doesn't actually do anything.

It used to be [wrongly used in place of the clang path, which I removed](https://github.com/apple/swift/pull/34437/files#diff-16566c91343a1660010c4c24db2b29a4fb1375cee942b57e04b194acce6c6861L105), so it is currently unused.

> So anyway, I agree, there shouldn't be four flags that are all for effectively the same purpose. I would like to split the adding of the new flag and removal of the old flags into separate PRs though.

Sounds good.

On replacing the separate clang path with a single toolchain path, we currently [juggle a couple different variables for compiling with clang to the target](https://github.com/apple/swift/blob/c6c88ce7140153ba828f580b821038c74a32cff7/stdlib/public/SwiftShims/CMakeLists.txt#L83), as opposed to the `--host-cc` variable for the clang that compiles for the host. We should make sure that keeps working.

I like the existing flexibility of having separate paths for the target clang toolchain and Swift toolchain, say if you're cross-compiling and have to use a separate C++ cross-compilation toolchain for your target, and I have used it that way in the past but I don't currently do so, as the Concurrency runtime has to built by the Swift-forked clang.

buttaface replied:
> the runtimes contain swiftcc/swift-asynccc, so it is unsafe to use an unmatched compiler to build those.

You are right that it is no longer something that can be done, though [I used to cross-compile the C++ files in the stdlib for Android with the Android NDK clang before](https://github.com/buttaface/swift-android-sdk/commit/bd7e6746cff3f9cc25b836ba481dc11e7ee62840#diff-76992bc322b65ab94f299c05ba3cbd8b3ab2e32ebfdde6db5adba70fa9f5a26aL132).

No need to add those additional build flags for me, as Android is already special-cased anyway.

etcwilde committed:
native-{}-tools-path -> custom-native-toolchain

This patch replaces the use of `native-(swift|clang|llvm)-tools-path`
with a single `custom-native-toolchain` flag in build-presets.ini.

etcwilde committed:
Add custom-native-toolchain tests

This adds the flag to the expected opts test and adds a test to ensure
it's passing the correct swiftc and clang paths to build-script cmake
products.

etcwilde commented:
@swift-ci please test

etcwilde commented on utils/build-presets.ini:
```diff
@@ -2386,11 +2386,9 @@ skip-test-cmark
-native-clang-tools-path=%(toolchain_path)s
+# Then set the paths to our native tools. If compiling against a toolchain, this
+# should point to the root of the toolchain.
+custom-native-toolchain=%(toolchain_path)s
```
I've replaced the native flags with this one. It should be a drop-in replacement assuming that the tools all live in the same place, so even if it's the `<root>/bin` dir, it's fine.

@buttaface I'm not sure if this affects Android or how you have that set up. While I'm pretty sure it should Just Work™, I'd like to actually verify it. Do you have instructions for doing this? I'm on macOS and have docker, but don't have any android devices.

buttaface replied:
I currently don't use any build presets, [specifying all the flags instead](https://github.com/buttaface/swift-android-sdk#building-the-android-sdks), so this change won't break anything for me.

etcwilde commented:
@swift-ci please test Linux

etcwilde commented:
@swift-ci please test macOS

buttaface commented on utils/swift_build_support/swift_build_support/products/swiftpm.py:
```diff
@@ -47,7 +47,10 @@ def run_bootstrap_script(self, action, host_target, additional_params=[]):
         toolchain_path = self.native_toolchain_path(host_target)
-        swiftc = os.path.join(toolchain_path, "bin", "swiftc")
+        if self.toolchain.swiftc is not None:
+            swiftc = self.toolchain.swiftc
```
We should switch over Python build products to always use this line in subsequent pulls, rather than doing the dance two lines below of adding and removing directories manually.

etcwilde replied:
First iteration, I did try going down the route of just using `swift.toolchain.swiftc`, but the Linux toolchain doesn't have a swiftc by default so we do need it to fall back on the `return self.install_toolchain_path(host_target)` in `native_toolchain_path` for normal builds.
Unfortunately, toolchains don't necessarily have a single root path, so I don't really know what the short-term right answer is.
Long-term, it seems like we need to discuss a refactor so that we don't need the `native_toolchain_path` at all since toolchains don't necessarily have one, and can pull what we need from `self.toolchain`, or fall back on the installed compiler (assuming it wasn't a Canadian cross compile).

gottesmm reviewed (approved):
I am fine with this. Test it carefully! Please be sure to remove the extra custom flags in the next PR.

etcwilde commented:
@swift-ci please test Linux

etcwilde commented:
@swift-ci please test macOS

etcwilde commented:
@swift-ci please test macOS

etcwilde commented:
@swift-ci please test macOS

etcwilde commented:
@swift-ci please test macOS

etcwilde commented:
@swift-ci please test macOS

etcwilde committed:
Pass swift and clang paths to build-script-impl

This patch updates the build-script-impl invocation to pass the paths to
the swift-tools and clang-tools from the toolchain to build-script-impl
if the tool is present. build-script-impl expects this to be to the path
containing the binary directly, so pulling the dirname from the absolute
path to the binary should be sufficient.

etcwilde commented:
@swift-ci please test

etcwilde commented:
Okay; so as far as I can tell `build-script-impl` has checks for the `NATIVE_SWIFT_TOOLS_PATH` variable that sets the swift compiler. If the var is set, it will set it to the path in that var. If it isn't set, it sets it to the thing we just built.

https://github.com/apple/swift/blob/1a9c951d202d324bd1c60dfa925cf331593caa27/utils/build-script-impl#L1731-L1741

Those get set by build-script here:
https://github.com/apple/swift/blob/a9527d4f2e1b2bd98cb2124466ec2af1f774953c/utils/swift_build_support/swift_build_support/build_script_invocation.py#L348-L359.

Since I removed the native swift tools path, build-script-impl was using the just-built 5.7 compiler to build instead of the compiler from the custom toolchain (5.5 from Xcode in this case). Then swift-driver used the toolchain to build (pulling in the 5.5 compiler). Since llbuild was built with 5.7, the modules were incompatible and the build failed.

I added

```python
        if args.custom_native_toolchain is not None:
             if self.toolchain.swiftc is not None:
                 swift_path = os.path.dirname(self.toolchain.swiftc)
                 impl_args += [
                     "--native-swift-tools-path=%s" % swift_path
                 ]
             if self.toolchain.clang is not None:
                 clang_path = os.path.dirname(self.toolchain.clang)
                 impl_args += [
                     "--native-clang-tools-path=%s" % clang_path
                 ]
```

so if we have a swift compiler, pull the path to that and use that as the `NATIVE_SWIFT_TOOLS_PATH` and if we have a clang compiler, set the path to that as the `NATIVE_CLANG_TOOLS_PATH`.

`NATIVE_LLVM_TOOLS_PATH` isn't hooked up to anything from what I can tell, so I'm not setting it. I'll try removing it entirely in a separate PR.

etcwilde commented:
Linux failure:

```
******************** TEST 'Swift-validation(linux-x86_64) :: BuildSystem/test_custom_toolchain.test' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp" && mkdir -p "/home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp"
: 'RUN: at line 2';   mkdir -p /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp
: 'RUN: at line 3';   touch /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp/cmake /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp/clang /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp/clang++ /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp/ninja /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp/swiftc
: 'RUN: at line 4';   SKIP_XCODE_VERSION_CHECK=1 SWIFT_BUILD_ROOT=/home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp /home/build-user/swift/utils/build-script --dry-run --skip-build --skip-early-swift-driver --custom-native-toolchain /home/build-user/build/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/BuildSystem/Output/test_custom_toolchain.test.tmp --cmark --static-curl --swiftpm 2>&1 | /usr/bin/python3.8 /home/build-user/swift/utils/PathSanitizingFileCheck --allow-unused-prefixes --sanitize BUILD_DIR=/home/build-user/build/buildbot_linux/swift-linux-x86_64 --sanitize SOURCE_DIR=/home/build-user/swift --use-filecheck /home/build-user/build/buildbot_linux/llvm-linux-x86_64/bin/FileCheck  /home/build-user/swift/validation-test/BuildSystem/test_custom_toolchain.test
--
Exit Code: 1

Command Output (stderr):
--
/home/build-user/swift/validation-test/BuildSystem/test_custom_toolchain.test:8:16: error: CHECK-LABEL: expected string not found in input
# CHECK-LABEL: --- Building cmark ---
               ^
<stdin>:1:1: note: scanning from here
[SOURCE_DIR/utils/build-script] NOTE: Using toolchain default
^
<stdin>:3:5: note: possible intended match here
+ /home/build-user/cmake/bootstrap --no-qt-gui -- -DCMAKE_USE_OPENSSL=OFF
    ^

Input file: <stdin>
Check file: /home/build-user/swift/validation-test/BuildSystem/test_custom_toolchain.test

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           1: [SOURCE_DIR/utils/build-script] NOTE: Using toolchain default 
label:8'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           2: + cmake --version 
label:8'0     ~~~~~~~~~~~~~~~~~~
           3: + /home/build-user/cmake/bootstrap --no-qt-gui -- -DCMAKE_USE_OPENSSL=OFF 
label:8'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:8'1         ?                                                                      possible intended match
           4: --------------------------------------------- 
label:8'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           5: CMake 3.19.6, Copyright 2000-2020 Kitware, Inc. and Contributors 
label:8'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           6: Found GNU toolchain 
label:8'0     ~~~~~~~~~~~~~~~~~~~~
           7: C compiler on this system is: gcc  
label:8'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           8: C++ compiler on this system is: g++  
label:8'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           .
           .
           .
>>>>>>

--

********************
```

etcwilde commented:
@swift-ci please test

etcwilde commented:
macOS failure:

```
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift-driver/Sources/SwiftDriverExecution/llbuild.swift:21:29: error: module compiled with Swift 5.7 cannot be imported by the Swift 5.5 compiler: /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/llbuild-macosx-x86_64/products/llbuildSwift/llbuildSwift.swiftmodule
@_implementationOnly import llbuildSwift
                            ^
```

still hitting it, so no dice. Though the above change was likely still necessary. Back to debugging.

etcwilde commented:
Okay, I've traced it back by comparing verbose build logs. It's the changes on which swift to use to build `swiftpm` with.

This is the original without my changes:
```
+ /Users/ewilde/Work/Cross-Compiling/sources/swiftpm/Utilities/bootstrap build --release --swiftc-path /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc --clang-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --cmake-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/cmake --ninja-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/ninja --build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/swiftpm-macosx-x86_64 --llbuild-build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/llbuild-macosx-x86_64 --reconfigure
--- bootstrap: note: Building swift-system
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The Swift compiler identification is Apple 5.7
```
It uses the swift 5.7 compiler from the build directory.

With my changes to use the toolchain `swiftc`, the swiftpm build-script bootstrap invocation is 

```
+ /Users/ewilde/Work/Cross-Compiling/sources/swiftpm/Utilities/bootstrap build --release --swiftc-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc --clang-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --cmake-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/cmake --ninja-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/ninja --build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/swiftpm-macosx-x86_64 --llbuild-build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/llbuild-macosx-x86_64 --reconfigure
--- bootstrap: note: Building swift-system
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The Swift compiler identification is Apple 5.5.2
```

The build that is failing is not using a custom native toolchain, it's using the normal macOS toolchain.
The toolchain has a swift compiler in it, specifically the swift compiler in Xcode. But in the normal case, it should be using the new compiler, not the one from the toolchain to build swiftpm. There isn't a great separation of host products and target products in build-script. I don't need swiftpm on the target so I can kind of ignore it, but it would be good to think about how we can layer things.

etcwilde commented:
Okay, I've traced it back by comparing verbose build logs. It's the changes on which swift to use to build `swiftpm` with.

This is the original without my changes:
```
+ /Users/ewilde/Work/Cross-Compiling/sources/swiftpm/Utilities/bootstrap build --release --swiftc-path /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc --clang-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --cmake-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/cmake --ninja-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/ninja --build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/swiftpm-macosx-x86_64 --llbuild-build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/llbuild-macosx-x86_64 --reconfigure
--- bootstrap: note: Building swift-system
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The Swift compiler identification is Apple 5.7
```
It uses the swift 5.7 compiler from the build directory.

With my changes to use the toolchain `swiftc`, the swiftpm build-script bootstrap invocation is 

```
+ /Users/ewilde/Work/Cross-Compiling/sources/swiftpm/Utilities/bootstrap build --release --swiftc-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc --clang-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --cmake-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/cmake --ninja-path /Applications/Xcode.app/Contents/Developer/usr/local/bin/ninja --build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/swiftpm-macosx-x86_64 --llbuild-build-dir /Users/ewilde/Work/Cross-Compiling/sources/build/ios_incremental/llbuild-macosx-x86_64 --reconfigure
--- bootstrap: note: Building swift-system
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The Swift compiler identification is Apple 5.5.2
```

The build that is failing is not using a custom native toolchain, it's using the normal macOS toolchain.
The toolchain has a swift compiler in it, specifically the swift compiler in Xcode. But in the normal case, it should be using the new compiler, not the one from the toolchain to build swiftpm. There isn't a great separation of host products and target products in build-script. I don't need swiftpm on the target so I can kind of ignore it, but it would be good to think about how we can layer things.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

buttaface commented:
> But in the normal case, it should be using the new compiler, not the one from the toolchain to build swiftpm. There isn't a great separation of host products and target products in build-script. I don't need swiftpm on the target so I can kind of ignore it, but it would be good to think about how we can layer things.

Unclear what you mean here: you are passing in the prebuilt Xcode toolchain to your new flag and having everything build with that instead of the freshly-built new compiler in the `build/` directory. Why would we build the corelibs with the Xcode toolchain but SPM with the new compiler?

I regularly [pass in a prebuilt linux compiler to the current `--native-swift-tools-path` when cross-compiling the toolchain from Linux x86_64 to Android](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L148), including SPM and sourcekit-lsp.

etcwilde commented:
> > But in the normal case, it should be using the new compiler, not the one from the toolchain to build swiftpm. There isn't a great separation of host products and target products in build-script. I don't need swiftpm on the target so I can kind of ignore it, but it would be good to think about how we can layer things.
> 
> Unclear what you mean here: you are passing in the prebuilt Xcode toolchain to your new flag and having everything build with that instead of the freshly-built new compiler in the `build/` directory. Why would we build the corelibs with the Xcode toolchain but SPM with the new compiler?
> 
> I regularly [pass in a prebuilt linux compiler to the current `--native-swift-tools-path` when cross-compiling the toolchain from Linux x86_64 to Android](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L148), including SPM and sourcekit-lsp.

Well, I thought I had an idea of what was going on, but in typing this up, I realized that I'm still missing a few pieces. Just going to write up what I've seen thus far.
I need to focus on this cmake invocation: 
```
env /Applications/Xcode.app/Contents/Developer/usr/local/bin/cmake \
  -G Ninja \
  -DCMAKE_C_COMPILER:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang \
  -DCMAKE_CXX_COMPILER:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ \
  -DCMAKE_Swift_COMPILER:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc \
  -DCMAKE_LIBTOOL:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool \
  -DCMAKE_AR:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar \
  -DCMAKE_RANLIB:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib \
  -DLLVM_VERSION_MAJOR:STRING=13 \
  -DLLVM_VERSION_MINOR:STRING=0 \
  -DLLVM_VERSION_PATCH:STRING=0 \
  -DCLANG_VERSION_MAJOR:STRING=13 \
  -DCLANG_VERSION_MINOR:STRING=0 \
  -DCLANG_VERSION_PATCH:STRING=0 \
  -DCMAKE_MAKE_PROGRAM=/Applications/Xcode.app/Contents/Developer/usr/local/bin/ninja \
  -DCMAKE_OSX_ARCHITECTURES=x86_64 \
  -DCMAKE_BUILD_TYPE:STRING=Release \
  -DCMAKE_C_COMPILER:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/llvm-macosx-x86_64/bin/clang \
  -DCMAKE_CXX_COMPILER:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/llvm-macosx-x86_64/bin/clang++ \
  -DCMAKE_INSTALL_PREFIX:PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/ \
  -DCMAKE_Swift_COMPILER:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/swift-macosx-x86_64/bin/swiftc \
  -DCMAKE_Swift_FLAGS:STRING= -module-cache-path "/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/llbuild-macosx-x86_64/module-cache" \
  -DLLBUILD_ENABLE_ASSERTIONS:BOOL=TRUE \
  -DLLBUILD_SUPPORT_BINDINGS:=Swift \
  -DLIT_EXECUTABLE:PATH=/Users/ewilde/Work/Cross-Compiling/sources/llvm-project/llvm/utils/lit/lit.py \
  -DFILECHECK_EXECUTABLE:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck \
  -DSWIFTC_EXECUTABLE:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/swift-macosx-x86_64/bin/swiftc \
  -DFOUNDATION_BUILD_DIR:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/foundation-macosx-x86_64 \
  -DLIBDISPATCH_BUILD_DIR:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/libdispatch-macosx-x86_64 \
  -DLIBDISPATCH_SOURCE_DIR:PATH=/Users/ewilde/Work/Cross-Compiling/sources/swift-corelibs-libdispatch \
  -Ddispatch_DIR:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/libdispatch-macosx-x86_64/cmake/modules \
  -DFoundation_DIR:PATH=/Users/ewilde/Work/Cross-Compiling/sources/build/buildbot_incremental/foundation-macosx-x86_64/cmake/modules \
  -DSQLite3_INCLUDE_DIR:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include \
  -DCMAKE_IGNORE_PATH=/usr/lib;/usr/local/lib;/lib \
  -DPKG_CONFIG_EXECUTABLE=/usr/bin/false \
  /Users/ewilde/Work/Cross-Compiling/sources/llbuild
```
It has the swift compiler set twice, and the second one is a 5.7 compiler, while the one used by the swift driver build is using a 5.5 compiler (from Xcode), so it fails to import the llbuild swift module.

Things to note, putting the `native-{swift|llvm|clang}-tools-path` variables back didn't fix the mismatched compilers in the build, so it's not that part. The problem is unrelated to the flag itself.



Anyway, since I already wrote it, here's where my initial thinking for dropping the swiftpm.py changes came from:

So this is as far as I can tell from comparing logs between the original and failing builds.
When building with the `buildbot_incremental` build preset on macOS, `self.toolchain` ends up being a `Darwin` toolchain, not a custom toolchain. Thus, when I do `self.toolchain.swiftc`, that will point to the the 5.5 compiler in the Xcode installed on the bot, not the 5.7 swiftc we just built.

https://github.com/apple/swift/blob/04256dcfb241f8c87de7e30a7e71cc9a85afd60c/utils/swift_build_support/swift_build_support/toolchain.py#L74-L90

In the original, swiftpm.py uses `self.native_toolchain_path` to get a path to the toolchain, then appends `bin` and `swiftc` to the file path to get the path to the swift compiler.

https://github.com/apple/swift/blob/04256dcfb241f8c87de7e30a7e71cc9a85afd60c/utils/swift_build_support/swift_build_support/products/swiftpm.py#L49-L50

Inside of the original `native_toolchain_path`, there's a check for `native_swift_tools_path`, which will be none, so it returns the `install_toolchain_path`. 

https://github.com/apple/swift/blob/04256dcfb241f8c87de7e30a7e71cc9a85afd60c/utils/swift_build_support/swift_build_support/products/product.py#L209-L213

So in the original 'normal' macOS build, we would use the swiftc from the toolchain to build a new swift compiler, but then use the new swift compiler to build swiftPM and friends. In the patch, I had changed this, so that swiftpm would check the the `toolchain.swiftc` and use that for the swiftpm/llbuild parts of the build. 

swift-driver is using the compiler from the toolchain, so it picks up the 5.5 compiler. Now that I'm typing this out, it seems like this should work since everything should be the 5.5 compiler coming from Xcode.

etcwilde commented:
I need to get back to this. I still need to dig into the swiftpm and llbuild build issue when using the toolchain compiler, but also need to get back to cross-compiling. I've created https://github.com/apple/swift/issues/59025 to track the missing pieces.

buttaface commented:
> It has the swift compiler set twice, and the second one is a 5.7 compiler, 

Yes, all the corelibs have these compilers set twice, once by the Python `build-script` and a second time by the bash `build-script-impl`, with the latter overriding the former. You can check this for a stock native build of these libraries too. At one point, somebody added a _third_ invocation, which I got removed.

> So in the original 'normal' macOS build, we would use the swiftc from the toolchain to build a new swift compiler, but then use the new swift compiler to build swiftPM and friends.

Yes, there is a distinction between what Swift components are allowed to be built with any prebuilt host Swift toolchain, currently only the early swift driver as the libswift portion of the compiler in `SwiftCompilerSources/` links against the freshly built stdlib, and what's built with the freshly built Swift compiler, such as SPM.

Now, if you're cross-compiling the toolchain, you obviously can't build anything with the freshly cross-compiled compiler, so you need to use either a freshly built native compiler or pass in a prebuilt native compiler _with the exact same version_. So you have to be careful, as while the host swift-driver can be built with a different Swift version, most of the cross-compiled corelibs and SPM cannot, so you may need to manage multiple prebuilt Swift toolchains _or_ only use one with the exact same version as the source, which is what I do when cross-compiling the toolchain for Android.
